validator/src/bootstrap.rs

```diff
@@ -616,629 +500,297 @@ mod without_incremental_snapshots {
-            for known_validator in known_validators {
-                cluster_info.get_snapshot_hash_for_node(known_validator, |snapshot_hashes| {
-                    for snapshot_hash in snapshot_hashes {
-                        known_snapshot_hashes.insert(*snapshot_hash);
-                    }
-                });
+/// Get the Slot and Hash of the local snapshot with the highest slot.  Can be either a full
+/// snapshot or an incremental snapshot.
+fn get_highest_local_snapshot_hash(
+    full_snapshot_archives_dir: impl AsRef<Path>,
+    incremental_snapshot_archives_dir: impl AsRef<Path>,
+    incremental_snapshot_fetch: bool,
+) -> Option<(Slot, Hash)> {
+    snapshot_utils::get_highest_full_snapshot_archive_info(full_snapshot_archives_dir).and_then(
+        |full_snapshot_info| {
+            if incremental_snapshot_fetch {
+                snapshot_utils::get_highest_incremental_snapshot_archive_info(
+                    incremental_snapshot_archives_dir,
+                    full_snapshot_info.slot(),
+                )
+                .map(|incremental_snapshot_info| {
+                    (
+                        incremental_snapshot_info.slot(),
+                        *incremental_snapshot_info.hash(),
+                    )
+                })
+            } else {
+                None
             }
-            Some(known_snapshot_hashes)
-        } else {
-            None
-        }
-    }
+            .or_else(|| Some((full_snapshot_info.slot(), *full_snapshot_info.hash())))
+        },
+    )
+}
 
-    /// Get the Slot and Hash of the local snapshot with the highest slot.
-    fn get_highest_local_snapshot_hash(
-        full_snapshot_archives_dir: impl AsRef<Path>,
-    ) -> Option<(Slot, Hash)> {
-        snapshot_utils::get_highest_full_snapshot_archive_info(full_snapshot_archives_dir)
-            .map(|full_snapshot_info| (full_snapshot_info.slot(), *full_snapshot_info.hash()))
+/// Get peer snapshot hashes
+///
+/// The result is a vector of peers with snapshot hashes that:
+/// 1. match a snapshot hash from the known validators
+/// 2. have the highest incremental snapshot slot
+/// 3. have the highest full snapshot slot of (2)
+fn get_peer_snapshot_hashes(
+    cluster_info: &ClusterInfo,
+    validator_config: &ValidatorConfig,
+    rpc_peers: &[ContactInfo],
+    incremental_snapshot_fetch: bool,
+) -> Vec<PeerSnapshotHash> {
+    let mut peer_snapshot_hashes =
+        get_eligible_peer_snapshot_hashes(cluster_info, rpc_peers, incremental_snapshot_fetch);
+    if validator_config.known_validators.is_some() {
+        let known_snapshot_hashes = get_snapshot_hashes_from_known_validators(
+            cluster_info,
+            validator_config,
+            incremental_snapshot_fetch,
+        );
+        retain_peer_snapshot_hashes_that_match_known_snapshot_hashes(
+            &known_snapshot_hashes,
+            &mut peer_snapshot_hashes,
+        );
     }
-}
+    if incremental_snapshot_fetch {
```

Comment:
This `if` shouldn't be needed, right? Assuming that `get_eligible_peer_snapshot_hashes()` and `get_snapshot_hashes_from_known_validators()` do not return any incremental snapshots if `incremental_snapshot_fetch == false`, then we're safe to keep the code as is.

Optionally could add an assert here instead. I.e. if `!incremental_snapshot_fetch` then `peer_snapshot_hashes` should not contain any incremental snapshot hashes.

(I understand that skipping the fn call is a potential perf improvement, but I would doubt the runtime of the fn is much at all; would be happy to see real-world data though!)